src/number_theory/legendre_symbol/add_character.lean

```diff
@@ -50,16 +52,54 @@ universes u v
 -- The domain of our additive characters
 variables (R : Type u) [add_monoid R]
 -- The target
-variables (R' : Type v) [monoid R']
+variables (R' : Type v) [comm_monoid R']
 
 /-- Define `add_char R R'` as `(multiplicative R) →* R'`.
 The definition works for an additive monoid `R` and a monoid `R'`,
 but we will restrict to the case that both are commutative rings below.
+We assume right away that `R'` is commutative, so that `add_char R R'` carries
+a structure of commutative monoid.
 The trivial additive character (sending everything to `1`) is `(1 : add_char R R').` -/
-abbreviation add_char : Type (max u v) := (multiplicative R) →* R'
+@[derive [comm_monoid, inhabited]]
+def add_char : Type (max u v) := (multiplicative R) →* R'
+
+instance : has_coe_to_fun (add_char R R') (λ x, (multiplicative R → R')) :=
+monoid_hom.has_coe_to_fun
+
+instance add_char.monoid_hom_class : monoid_hom_class (add_char R R') (multiplicative R) R' :=
+monoid_hom.monoid_hom_class
```

Comment:
`monoid_hom_class` extends `fun_like`, and we already have [fun_like.has_coe_to_fun](https://leanprover-community.github.io/mathlib_docs/data/fun_like/basic.html#fun_like.has_coe_to_fun).

Reply:
If I comment out the `has_coe_to_fun` instance, I get an error message in line 260 (on `coe_fn`) complaining that the type class instance `has_coe_to_fun (add_char (zmod n) C) (λ (ᾰ : add_char (zmod n) C), multiplicative (zmod n) → C)` cannot be synthesized. So it looks like the instance is still needed.

Reply:
Pretty weird;
```
instance : has_coe_to_fun (add_char R R') (λ x, (multiplicative R → R')) :=
fun_like.has_coe_to_fun
```
after the second instance works, and `fun_like.has_coe_to_fun` is an instance. Why doesn't `by apply_instance` work?

Reply:
Perhaps this is because `fun_like` applies to the `hom_class` and `has_coe_to_fun` to the `hom`s themselves? But I don't really know.